### DISCUSSION TITLE
can API be disabled to outside while UI is enabled?

### DISCUSSION THREAD

**Discussion by tooptoop4** [argoproj/argo-workflows#13817] [Q&A] [answered]
i imagine UI needs to call API so this would only be possible if they could run on different ports? then Ingress only opened on the UI port

this is related to https://github.com/argoproj/argo-workflows/issues/11391

**Comment by agilgur5** [comment, marked as answer]
> i imagine UI needs to call API

Correct, so there's no way to truly distinguish a browser client from any other API client; any attempt to distinguish would be spoofable.

> then Ingress only opened on the UI port

Then the UI would be unable to access the API. Remember the UI is running in your browser, so your browser must be able to access the API.

Not really an Argo specific question

**Reply by tooptoop4** [reply]
it could be server side UI code?

**Reply by agilgur5** [reply]
I don't know what you mean by that. The "server side UI code" _is_ the API. Again, the UI is just another client of the API, one that happens to run in a browser.
If you mean rendering on the server, other putting much more load on the server, being non-trivial in Go, and being a massive refactor and architectural shift, that also does not truly change access patterns, as those renderable endpoints must be reachable by a browser.
Again, this isn't Argo specific.

**Reply by tooptoop4** [reply]
browser > hit 443 ingress alb of UI (this port maps to pod port of 8080) > the underlying argo-server pod of this UI has received this request on internal 8080 port that pod exposes > the argo-server pod internally talks to API port 80 of argo-server pod via local port or k8s service (not ingress) > argo-server pod 8080 is then able to provide html response. port 80 is never mapped to an ingress

**Reply by agilgur5** [reply]
> argo-server pod 8080 is then able to provide html response

This is effectively the same as server side rendering, so it has the same issues as I wrote above. An HTML response is basically just an obfuscation of the API response.

The proper way to handle this is via auth and firewalling (NetworkPolicy, VPN, etc). The API should only be accessible to authenticated users from within a VPN.

Your suggestion is also very similar to how server-side virtual desktops ([VDIs](https://azure.microsoft.com/en-us/resources/cloud-computing-dictionary/what-is-virtual-desktop-infrastructure-vdi)) and similar type tools like [remote browser isolation](https://www.cloudflare.com/zero-trust/products/browser-isolation/) (RBI) work. Pretty out-of-scope.

**Reply by tooptoop4** [reply]
grafana uses some host allow list https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-request-security/

**Reply by agilgur5** [reply]
That's for server-side egress in Grafana, not ingress. You can also achieve the same via NetworkPolicies
